### PR TITLE
Attribute whitelist that allows negative attributes

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature:** This PR implements the feature request detailed and discussed in issue #4682 

## Feature Details
Adds a hardcoded whitelist of attributes that should be allowed to be negative on a ship. To start, I've added the repair cost attributes that as of #4932 have functionality when negative. (Ref: #4525 from where I copied over some of this code.)

## UI Screenshots
N/A

## Usage Examples
N/A

## Testing Done
I put `"shield energy" -1` on some cooling ducts and installed a bunch.  The recharging shields energy cell of the energy/heat table became positive, showing that shield energy was now at a net negative; previously, this would not be possible, as the outfitter would block you from having negative attributes. And thanks to #4932, being attacked meant my energy production shot up.

## Performance Impact
N/A

### CODE REVIEW

**Comment by Zitchas** [comment]
I know the initially impression of taking hits *generating* energy sounds odd initially, but for, say, a water cooling system, just imagine having a "steam turbine" outfit that could be added on to it. When the system produces a lot of heat, it runs the turbine and generates power! Theoretically, anything that produces heat could be hooked up to a turbine system.

**Comment by tehhowch** [comment]
I'm not a fan of a user-configured whitelist, as it's extremely easy to "break" the game. Consider what happens when the whitelist include things like "outfit space", any of the ammo capacity attributes, etc. 

While offering customization of the game experience is something that plugins should be able to do, fundamentally violating the core constructs of the game is not.

**Comment by Amazinite** [comment]
If somebody really wants to break the game, I don't see a reason why we should stop them. I can hard code the whitelist in Outfit.cpp though; any use case I can imagine for an easily customizable whitelist all involved being combined with #4528, but with the suggestion of that being expanded to allow basically using a ConditionSet with attributes, anything that would be done with a single attribute that can be negative could also be done with a comparison of two attributes.

e.g. say we have an alien civilization with outfits that have a "taboo" attribute; some outfits are taboo to have (positive values), while others are seen as good to have (negative values), then missions could be offered based off of your net taboo value. Without allowing negative attributes though, the good outfits could instead have a different attribute (let's say "not taboo" because I can't think of anything else), then to do the same mission offering you'd just do `attribute "not taboo" >= "taboo"` or what have you.

**Comment by Amazinite** [comment]
@tehhowch Everything look good here now?
(Note to self: #4519 needs minimum values of -0.99, #4525 needs minimum values of -1.)

File: source/Outfit.cpp [hunk 1]
```diff
@@ -201,10 +216,21 @@ int Outfit::CanAdd(const Outfit &other, int count) const
 {
 	for(const auto &at : other.attributes)
 	{
+		// The minimum allowed value of most attributes is 0. Some attributes
+		// have special functionality when negative, though, and are therefore
+		// allowed to have values less than 0.
+		double minimum = 0.;
+		if(WHITELIST.count(at.first))
```
Comment (tehhowch) [inline source/Outfit.cpp:223]: Don't double-scan the map, by using count and then find.

Comment (Amazinite) [reply source/Outfit.cpp:223]: So instead 
```
auto it = WHITELIST.find(at.first);
if(it != WHITELIST.end()) {
	minimum = it->second;
	...
}
```
?

**Pull request merged**